---
title: "Aging"
author: "Philipp Ross"
date: "2018-10-06"
output: workflowr::wflow_html
editor_options: 
  chunk_output_type: console
---

We woud like to categorize the genes into parent and offspring pairs. In order to do this, we need to estimate the earliest common ancestor with an orthologue of these genes.

```{r}
library(tidyverse)
library(cowplot)
library(org.Hs.eg.db)
```

## Ortholog retrieval

We can use biomart to download several orthologs and find out how far back we can date these genes.

```{r}
conservative_gene_list <- readr::read_tsv("../output/thirdset/conservative_reciprocal_best_hits.tsv",col_names=T)$query
standard_hits <- readr::read_tsv("../output/thirdset/standard_hits.tsv",col_names=T)
standard_gene_list <- dplyr::union(standard_hits$query,standard_hits$subject)

human      <- biomaRt::useMart("ensembl", dataset = "hsapiens_gene_ensembl")
chimp      <- biomaRt::useMart("ensembl", dataset = "ptroglodytes_gene_ensembl")
orangutan  <- biomaRt::useMart("ensembl", dataset = "pabelii_gene_ensembl")
rhesus     <- biomaRt::useMart("ensembl", dataset = "mmulatta_gene_ensembl")
marmoset   <- biomaRt::useMart("ensembl", dataset = "cjacchus_gene_ensembl")
mouse      <- biomaRt::useMart("ensembl", dataset = "mmusculus_gene_ensembl")
guinea_pig <- biomaRt::useMart("ensembl", dataset = "cporcellus_gene_ensembl")
dog        <- biomaRt::useMart("ensembl", dataset = "cfamiliaris_gene_ensembl")
cow        <- biomaRt::useMart("ensembl", dataset = "btaurus_gene_ensembl")
armadillo  <- biomaRt::useMart("ensembl", dataset = "dnovemcinctus_gene_ensembl")
opossum    <- biomaRt::useMart("ensembl", dataset = "mdomestica_gene_ensembl")
platypus   <- biomaRt::useMart("ensembl", dataset = "oanatinus_gene_ensembl")
chicken    <- biomaRt::useMart("ensembl", dataset = "ggallus_gene_ensembl")
lizard     <- biomaRt::useMart("ensembl", dataset = "acarolinensis_gene_ensembl")
frog       <- biomaRt::useMart("ensembl", dataset = "xtropicalis_gene_ensembl")
coelacanth <- biomaRt::useMart("ensembl", dataset = "lchalumnae_gene_ensembl")
fugu       <- biomaRt::useMart("ensembl", dataset = "trubripes_gene_ensembl")
zebrafish  <- biomaRt::useMart("ensembl", dataset = "drerio_gene_ensembl")
hagfish    <- biomaRt::useMart("ensembl", dataset = "eburgeri_gene_ensembl")
lamprey    <- biomaRt::useMart("ensembl", dataset = "pmarinus_gene_ensembl")
worm       <- biomaRt::useMart("ensembl", dataset = "celegans_gene_ensembl")
fly        <- biomaRt::useMart("ensembl", dataset = "dmelanogaster_gene_ensembl")
yeast      <- biomaRt::useMart("ensembl", dataset = "scerevisiae_gene_ensembl")

species_lookup_table <- c("human","chimp","orangutan","rhesus","marmoset","mouse",
                  "guinea_pig","dog","cow","armadillo","oppossum","platypus",
                  "chicken","lizard","frog","coelacanth","fugu","zebrafish",
                  "hagfish","lamprey","worm","fly","yeast")

#species_lookup_table <- c("human","chimp","orangutan","rhesus","marmoset","mouse",
#                  "guinea_pig","dog","cow","armadillo","oppossum","platypus",
#                  "chicken","lizard","frog","coelacanth","fugu","zebrafish",
#                  "worm","fly","yeast")

branch_lookup_table <- c("16","15","14","13","12","11","11","10","10","9","8",
                         "7","6","6","5","4","3","3","2","2","1","1","0")

#branch_lookup_table <- c("15","14","13","12","11","10","10","9","9","8",
#                         "7","6","5","5","4","3","2","2","1","1","0")

branch_lengths <- tibble(branch=c(16:0),branch_length=c(6,6,7,12,18,27,30,60,60,90,60,30,50,100,250,500,700))

dbs <- c(human,chimp,orangutan,rhesus,marmoset,mouse,guinea_pig,dog,cow,
         armadillo,opossum,platypus,chicken,lizard,frog,coelacanth,fugu,
         zebrafish,hagfish,lamprey,worm,fly,yeast)

#dbs <- c(human,chimp,orangutan,rhesus,marmoset,mouse,guinea_pig,dog,cow,
#         armadillo,opossum,platypus,chicken,lizard,frog,coelacanth,fugu,
#         zebrafish,worm,fly,yeast)
```

```{r}
# Function to retrieve homologs from non-human species and return the:
#   human gene id
#   alternative organism gene id
#   alternative organism name
#   the branch number corresponding to the last common ancestor of the two homologs
retrieve_orhologs <- function(species_lookup_table, branch_lookup_table, dbs, gene_list) {
  out <- tibble::tibble(human_id=character(),
                        ortholog_id=character(),
                        organism=character())
  
  for (i in 2:length(species_lookup_table)) {
      
      result <- biomaRt::getLDS(
        attributes = c("ensembl_gene_id"), 
        filters = "ensembl_gene_id", 
        values = gene_list,
        mart = human,
        attributesL = c("ensembl_gene_id"), 
        martL = dbs[[i]])
      
      result <- dplyr::rename(result,human_id=Gene.stable.ID,ortholog_id=Gene.stable.ID.1) %>%
        dplyr::mutate(organism=species_lookup_table[i],branch=branch_lookup_table[i]) %>%
        tibble::as_tibble()
      
      out <- dplyr::bind_rows(out,result)
  }
  return(out)
}
```

### Conservative set orthologs

```{r}
conservative_orthos <- retrieve_orhologs(species_lookup_table,branch_lookup_table,dbs,conservative_gene_list)

conservative_in_humans <- length(unique(conservative_gene_list)) - length(unique(conservative_orthos$human_id))
```

### Standard set orthologs

```{r}
standard_orthos <- retrieve_orhologs(species_lookup_table,branch_lookup_table,dbs,standard_gene_list)

standard_in_humans <- length(unique(standard_gene_list)) - length(unique(standard_orthos$human_id))
```

## Parent-offspring assignments

### Conservative set

```{r}
conservative_rbh <- readr::read_tsv("../output/thirdset/conservative_reciprocal_best_hits.tsv",col_names=T)
human_orthos <- conservative_orthos %>% 
  dplyr::group_by(human_id) %>% 
  dplyr::summarize(oldest=(min(as.integer(branch)))) %>%
  dplyr::ungroup()  

tmp1 <- human_orthos %>%
  dplyr::inner_join(conservative_rbh, by=c("human_id"="query")) %>%
  dplyr::rename(query=human_id,query_branch=oldest)

tmp2 <- human_orthos %>%
  dplyr::inner_join(conservative_rbh, by=c("human_id"="subject")) %>%
  dplyr::rename(subject=human_id,subject_branch=oldest)

conservative_potbl <- dplyr::inner_join(tmp1,tmp2)
rm(tmp1,tmp2)

conservative_potbl$query_relation <- apply(conservative_potbl, 1, function(row) {ifelse(row[[2]] == row[[4]],"same",ifelse(row[[2]] < row[[4]],"parent","offspring"))})
conservative_potbl$subject_relation <- apply(conservative_potbl, 1, function(row) {ifelse(row[[2]] == row[[4]],"same",ifelse(row[[2]] > row[[4]],"parent","offspring"))})

# some genes have more than one annotated gene name? Just keep one.
tmp <- dplyr::inner_join(as.data.frame(org.Hs.egGENENAME),as.data.frame(org.Hs.egENSEMBL)) %>%
  dplyr::select(ensembl_id,gene_name) %>%
  tibble::as_tibble() 

gtbl <- list()

for (i in 1:nrow(tmp)) {
  row <- tmp[i,]
  ifelse(is.null(gtbl[[row$ensembl_id]]),
         gtbl[[row$ensembl_id]] <- row$gene_name,
         next())
}

gtbl <- tibble::tibble(ensembl_id=names(gtbl),gene_name=as.vector(unlist(gtbl)))

conservative_potbl <- dplyr::inner_join(conservative_potbl,gtbl,by=c("query"="ensembl_id")) %>%
  dplyr::rename(query_gene_name=gene_name)

conservative_potbl <- dplyr::inner_join(conservative_potbl,gtbl,by=c("subject"="ensembl_id")) %>%
  dplyr::rename(subject_gene_name=gene_name)

query   <- conservative_potbl[,c(1,2,5)] %>% dplyr::rename(gene_id=query,branch=query_branch,relation=query_relation)
subject <- conservative_potbl[,c(3,4,6)] %>% dplyr::rename(gene_id=subject,branch=subject_branch,relation=subject_relation)
conservative_plot_this <- dplyr::bind_rows(query,subject) %>% dplyr::distinct()

conservative_normalized_events <- conservative_plot_this %>% 
  dplyr::group_by(gene_id) %>% 
  dplyr::group_by(branch,relation) %>%
  dplyr::summarize(`Number of genes`=n()) %>%
  dplyr::ungroup() %>%
  tidyr::spread(key=relation,value=`Number of genes`,fill=0) %>%
  dplyr::mutate(`Number of Events`=offspring+(same/2)) %>%
  dplyr::inner_join(branch_lengths) %>%
  dplyr::bind_rows(.,tibble::tibble(branch=16,offspring=conservative_in_humans,parent=0,same=0,`Number of Events`=conservative_in_humans,branch_length=6)) %>%
  dplyr::mutate(`Events per million years`=`Number of Events`/branch_length)

readr::write_tsv(x=conservative_normalized_events,"../output/aging/conservative_normalized_events.tsv")

g <- conservative_normalized_events %>%
  ggplot(aes(x=branch,y=`Events per million years`)) + 
  geom_bar(stat="identity") +
  xlab("Oldest Branch") + 
  ylab("Events per Million Years") +
  ggtitle("Conservative") + 
  scale_x_continuous(breaks=0:16,labels=c("0","1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16"))
 
cowplot::save_plot("../output/aging/conservative_events_over_time.svg",g)
cowplot::save_plot("../output/aging/conservative_events_over_time.png",g)
print(g)
```

### Standard set

```{r}
standard_hits <- readr::read_tsv("../output/thirdset/standard_hits.tsv",col_names=T)
human_orthos <- standard_orthos %>% 
  dplyr::group_by(human_id) %>% 
  dplyr::summarize(oldest=(min(as.integer(branch)))) %>%
  dplyr::ungroup()  

tmp1 <- human_orthos %>%
  dplyr::inner_join(standard_hits, by=c("human_id"="query")) %>%
  dplyr::rename(query=human_id,query_branch=oldest)

tmp2 <- human_orthos %>%
  dplyr::inner_join(standard_hits, by=c("human_id"="subject")) %>%
  dplyr::rename(subject=human_id,subject_branch=oldest)

standard_potbl <- dplyr::inner_join(tmp1,tmp2)
rm(tmp1,tmp2)

standard_potbl$query_relation <- apply(standard_potbl, 1, function(row) {ifelse(row[[2]] == row[[4]],"same",ifelse(row[[2]] < row[[4]],"parent","offspring"))})
standard_potbl$subject_relation <- apply(standard_potbl, 1, function(row) {ifelse(row[[2]] == row[[4]],"same",ifelse(row[[2]] > row[[4]],"parent","offspring"))})

standard_potbl <- dplyr::inner_join(standard_potbl,gtbl,by=c("query"="ensembl_id")) %>%
  dplyr::rename(query_gene_name=gene_name)

standard_potbl <- dplyr::inner_join(standard_potbl,gtbl,by=c("subject"="ensembl_id")) %>%
  dplyr::rename(subject_gene_name=gene_name)

query   <- standard_potbl[,c(1,2,5)] %>% dplyr::rename(gene_id=query,branch=query_branch,relation=query_relation)
subject <- standard_potbl[,c(3,4,6)] %>% dplyr::rename(gene_id=subject,branch=subject_branch,relation=subject_relation)
standard_plot_this <- dplyr::bind_rows(query,subject)

offspring <- integer(length=16)
parent <- integer(length=16)
same <- integer(length=16)
hits <- list()
count <- 0
# loop through table lacking reciprocal best hits making sure we
# count cases of A to B and B to A as one event, but
# cases of A to B and A to C as two events and A to B and B to C as two events
for (i in 1:nrow(standard_potbl)) {
  row <- standard_potbl[i,]
  qbranch <- row[[2]] + 1
  sbranch <- row[[4]] + 1
  qrelation <- row[[5]]
  srelation <- row[[6]]
  if (is.null(hits[[row$query]])) {
    if (is.null(hits[[row$subject]])) {
      count <- count + 1
      hits[[row$query]] <- row$subject
      if(qrelation==srelation) {
        same[qbranch] <- same[qbranch] + 1
      } else if(qrelation=="offspring") {
        offspring[qbranch] <- offspring[qbranch] + 1
        parent[sbranch] <- parent[sbranch] + 1
      } else if(qrelation=="parent") {
        offspring[sbranch] <- offspring[sbranch] + 1
        parent[qbranch] <- parent[qbranch] + 1
      }
      next()
    } else if (row$query %in% hits[[row$subject]]) {
      hits[[row$query]] <- row$subject
      next()
    } else {
      count <- count + 1
      hits[[row$query]] <- row$subject
      if(qrelation==srelation) {
        same[qbranch] <- same[qbranch] + 1
      } else if(qrelation=="offspring") {
        offspring[qbranch] <- offspring[qbranch] + 1
        parent[sbranch] <- parent[sbranch] + 1
      } else if(qrelation=="parent") {
        offspring[sbranch] <- offspring[sbranch] + 1
        parent[qbranch] <- parent[qbranch] + 1
      }
    }
    next()
  } else if (row$subject %in% hits[[row$query]]) {
    next()
  } else {
    count <- count + 1
    hits[[row$query]] <- c(hits[[row$query]],row$subject)
      if(qrelation==srelation) {
        same[qbranch] <- same[qbranch] + 1
      } else if(qrelation=="offspring") {
        offspring[qbranch] <- offspring[qbranch] + 1
        parent[sbranch] <- parent[sbranch] + 1
      } else if(qrelation=="parent") {
        offspring[sbranch] <- offspring[sbranch] + 1
        parent[qbranch] <- parent[qbranch] + 1
      }
  }
  next()
}

standard_normalized_events <- tibble::tibble(branch=0:15,
                                     offspring=offspring,
                                     parent=parent,
                                     same=same,
                                     `Number of Events`=offspring+same) %>%
  dplyr::inner_join(branch_lengths) %>%
  dplyr::bind_rows(tibble::tibble(branch=16,offspring=standard_in_humans,parent=0,same=0,`Number of Events`=standard_in_humans,branch_length=6)) %>%
  dplyr::mutate(`Events per million years`=`Number of Events`/branch_length)

readr::write_tsv(x=standard_normalized_events,"../output/aging/standard_normalized_events.tsv")

g <- standard_normalized_events %>%
  ggplot(aes(x=branch,y=`Events per million years`)) + 
  geom_bar(stat="identity") +
  xlab("Oldest Branch") + 
  ylab("Events per Million Years") +
  ggtitle("Standard") + 
  scale_x_continuous(breaks=0:16,labels=c("0","1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16"))
 
cowplot::save_plot("../output/aging/standard_events_over_time.svg",g)
cowplot::save_plot("../output/aging/standard_events_over_time.png",g)
print(g)
```

## Tissue expression analysis

### What tissues are these genes expressed in?

#### Conservative set

```{r,fig.height=8}
tpm_threshold <- 2
relationships <- c("same","parent","offspring")
gtex <- readr::read_tsv("../data/GTEx_Analysis_2016-01-15_v7_RNASeQCv1.1.8_gene_median_tpm.gct",skip=2,col_names=T) %>%
  dplyr::select(-Description) %>%
  tidyr::gather(key = tissue, value = exp, -gene_id)

# remove annoying ".#" parts of each gene ID
gtex$gene_id <- sapply(stringr::str_split(gtex$gene_id,"[.]"), function(x) {x[1]})

fgtex <- dplyr::inner_join(gtex,conservative_plot_this) %>%
  dplyr::distinct()

tissue_expression <- fgtex %>%
  dplyr::filter(exp > tpm_threshold) %>%
  dplyr::group_by(tissue) %>%
  dplyr::summarise(n=n()) %>%
  dplyr::arrange(desc(n))

g <- ggplot(tissue_expression, aes(x=reorder(tissue, n),y=n)) +
  geom_bar(stat="identity") +
  coord_flip() +
  ylab("Number of genes") +
  xlab("") + 
  ggtitle("Total")

cowplot::save_plot(paste0("../output/aging/conservative_total_tissue_expression_barplot.png"),plot=g)

print(g)

for (r in relationships) {
  tissue_expression <- fgtex %>%
    dplyr::filter(relation==r) %>%
    dplyr::filter(exp > tpm_threshold) %>%
    dplyr::group_by(tissue) %>%
    dplyr::summarise(n=n()) %>%
    dplyr::arrange(desc(n))
  
  g <- ggplot(tissue_expression, aes(x=reorder(tissue, n),y=n)) +
    geom_bar(stat="identity") +
    coord_flip() +
    ylab("Number of genes") +
    xlab("") + 
    ggtitle(r)
  
  cowplot::save_plot(paste0("../output/aging/conservative_",r,"_tissue_expression_barplot.png"),plot=g)
  
  print(g)
}
```

#### Standard set

```{r,fig.height=8}
fgtex <- dplyr::inner_join(gtex,standard_plot_this) %>%
  dplyr::distinct()

tissue_expression <- fgtex %>%
  dplyr::filter(exp > tpm_threshold) %>%
  dplyr::group_by(tissue) %>%
  dplyr::summarise(n=n()) %>%
  dplyr::arrange(desc(n))

g <- ggplot(tissue_expression, aes(x=reorder(tissue, n),y=n)) +
  geom_bar(stat="identity") +
  coord_flip() +
  ylab("Number of genes") +
  xlab("") + 
  ggtitle("Total")

cowplot::save_plot(paste0("../output/aging/standard_total_tissue_expression_barplot.png"),plot=g)

print(g)

for (r in relationships) {
  tissue_expression <- fgtex %>%
    dplyr::filter(relation==r) %>%
    dplyr::filter(exp > tpm_threshold) %>%
    dplyr::group_by(tissue) %>%
    dplyr::summarise(n=n()) %>%
    dplyr::arrange(desc(n))
  
  g <- ggplot(tissue_expression, aes(x=reorder(tissue, n),y=n)) +
    geom_bar(stat="identity") +
    coord_flip() +
    ylab("Number of genes") +
    xlab("") + 
    ggtitle(r)
  
  cowplot::save_plot(paste0("../output/aging/standard_",r,"_tissue_expression_barplot.png"),plot=g)
  
  print(g)
}
```

### How many tissues are most genes expressed in?

#### Conservative set

```{r}
fgtex <- dplyr::inner_join(gtex,conservative_plot_this) %>%
  dplyr::distinct()

# in the standard set, some genes have more than one relationship to other genes
# i.e. Gene A is same to Gene B, but the parent of Gene C
# the last line, dplyr::mutate(n=ifelse(n>53,n/2,n)), takes this into account
# When they have two, the number is always double what it should be
gene_expression <- fgtex %>%
  dplyr::group_by(gene_id) %>%
  dplyr::summarise(n = sum(exp > tpm_threshold)) %>%
  dplyr::mutate(n=ifelse(n>53,n/2,n)) 

g <- ggplot(gene_expression, aes(x=n)) +
  geom_histogram(color="grey70") +
  xlab("Number of Tissues") +
  ylab("Number of Genes") + 
  ggtitle("Total")

cowplot::save_plot("../output/aging/conservative_total_gene_expression_distribution.png",plot=g)

print(g)

for (r in relationships) {
  gene_expression <- gtex %>%
    dplyr::inner_join(conservative_plot_this) %>%
    dplyr::distinct() %>%
    dplyr::filter(relation==r) %>%
    dplyr::group_by(gene_id) %>%
    dplyr::summarise(n = sum(exp > tpm_threshold))
  
  g <- ggplot(gene_expression, aes(x=n)) +
    geom_histogram(color="grey70") +
    xlab("Number of Tissues") +
    ylab("Number of Genes") + 
    ggtitle(r)
  
  cowplot::save_plot(paste0("../output/aging/conservative_",r,"_gene_expression_distribution.png"),plot=g)
  
  print(g)
}
```

```{r}
g <- fgtex %>%
  ggplot(aes(x=relation,y=log2(exp+1))) +
  geom_boxplot() +
  xlab("Relationship") +
  ylab("Log2 Expression")
  
cowplot::save_plot("../output/aging/conservative_gene_expression_comparison.png",plot=g)
  
print(g)
```

#### Standard set

```{r}
fgtex <- dplyr::inner_join(gtex,standard_plot_this) %>%
  dplyr::distinct()

# in the standard set, some genes have more than one relationship to other genes
# i.e. Gene A is same to Gene B, but the parent of Gene C
# the last line, dplyr::mutate(n=ifelse(n>53,n/2,n)), takes this into account
# When they have two, the number is always double what it should be
gene_expression <- fgtex %>%
  dplyr::group_by(gene_id) %>%
  dplyr::summarise(n = sum(exp > tpm_threshold)) %>%
  dplyr::mutate(n=ifelse(n>53,n/2,n)) 

g <- ggplot(gene_expression, aes(x=n)) +
  geom_histogram(color="grey70") +
  xlab("Number of Tissues") +
  ylab("Number of Genes") + 
  ggtitle("Total")

cowplot::save_plot("../output/aging/standard_total_gene_expression_distribution.png",plot=g)

print(g)

for (r in relationships) {
  gene_expression <- fgtex %>%
    dplyr::filter(relation==r) %>%
    dplyr::group_by(gene_id) %>%
    dplyr::summarise(n = sum(exp > tpm_threshold))
  
  g <- ggplot(gene_expression, aes(x=n)) +
    geom_histogram(color="grey70") +
    xlab("Number of Tissues") +
    ylab("Number of Genes") + 
    ggtitle(r)
  
  cowplot::save_plot(paste0("../output/aging/standard_",r,"_gene_expression_distribution.png"),plot=g)
  
  print(g)
}
```

```{r}
g <- fgtex %>%
  ggplot(aes(x=relation,y=log2(exp+1))) +
  geom_boxplot() +
  xlab("Relationship") +
  ylab("Log2 Expression")
  
cowplot::save_plot("../output/aging/standard_gene_expression_comparison.png",plot=g)
  
print(g)
```

## Save results to file

```{r}
one <- gtex %>% 
  dplyr::filter(exp > tpm_threshold) %>% 
  dplyr::group_by(gene_id) %>% 
  dplyr::summarise(n=n()) %>% 
  dplyr::inner_join(conservative_potbl,by=c("gene_id"="query")) %>% 
  dplyr::rename(query_tissues=n)

out <- gtex %>% 
  dplyr::filter(exp > tpm_threshold) %>% 
  dplyr::group_by(gene_id) %>% 
  dplyr::summarise(n=n()) %>% 
  dplyr::inner_join(one,by=c("gene_id"="subject")) %>% 
  dplyr::rename(query_id=gene_id,subject_id=gene_id.y,subject_tissues=n) %>%
  dplyr::select(query_id,query_tissues,query_branch,query_relation,query_gene_name,
                subject_id,subject_tissues,subject_branch,subject_relation,subject_gene_name)

readr::write_tsv(x=out,path="../output/aging/conservative_data_table.tsv")

one <- gtex %>% 
  dplyr::filter(exp > tpm_threshold) %>% 
  dplyr::group_by(gene_id) %>% 
  dplyr::summarise(n=n()) %>% 
  dplyr::inner_join(standard_potbl,by=c("gene_id"="query")) %>% 
  dplyr::rename(query_tissues=n)

out <- gtex %>% 
  dplyr::filter(exp > tpm_threshold) %>% 
  dplyr::group_by(gene_id) %>% 
  dplyr::summarise(n=n()) %>% 
  dplyr::inner_join(one,by=c("gene_id"="subject")) %>% 
  dplyr::rename(query_id=gene_id,subject_id=gene_id.y,subject_tissues=n) %>%
  dplyr::select(query_id,query_tissues,query_branch,query_relation,query_gene_name,
                subject_id,subject_tissues,subject_branch,subject_relation,subject_gene_name)

readr::write_tsv(x=out,path="../output/aging/standard_data_table.tsv")
```
