---
title: "Aging"
author: "Philipp Ross"
date: "2018-10-06"
output: workflowr::wflow_html
editor_options: 
  chunk_output_type: console
---

We woud like to categorize the genes into parent and offspring pairs. In order to do this, we need to estimate the earliest common ancestor with an orthologue of these genes.

```{r}
library(tidyverse)
library(cowplot)
```

## Ortholog retrieval

We can use biomart to download several orthologs and find out how far back we can date these genes.

```{r}
conservative_gene_list <- readr::read_tsv("../output/thirdset/conservative_reciprocal_best_hits.tsv",col_names=T)$query
standard_gene_list <- readr::read_tsv("../output/thirdset/standard_hits.tsv",col_names=T)$query

human      <- biomaRt::useMart("ensembl", dataset = "hsapiens_gene_ensembl")
chimp      <- biomaRt::useMart("ensembl", dataset = "ptroglodytes_gene_ensembl")
orangutan  <- biomaRt::useMart("ensembl", dataset = "pabelii_gene_ensembl")
rhesus     <- biomaRt::useMart("ensembl", dataset = "mmulatta_gene_ensembl")
marmoset   <- biomaRt::useMart("ensembl", dataset = "cjacchus_gene_ensembl")
mouse      <- biomaRt::useMart("ensembl", dataset = "mmusculus_gene_ensembl")
guinea_pig <- biomaRt::useMart("ensembl", dataset = "cporcellus_gene_ensembl")
dog        <- biomaRt::useMart("ensembl", dataset = "cfamiliaris_gene_ensembl")
cow        <- biomaRt::useMart("ensembl", dataset = "btaurus_gene_ensembl")
armadillo  <- biomaRt::useMart("ensembl", dataset = "dnovemcinctus_gene_ensembl")
opossum    <- biomaRt::useMart("ensembl", dataset = "mdomestica_gene_ensembl")
platypus   <- biomaRt::useMart("ensembl", dataset = "oanatinus_gene_ensembl")
chicken    <- biomaRt::useMart("ensembl", dataset = "ggallus_gene_ensembl")
lizard     <- biomaRt::useMart("ensembl", dataset = "acarolinensis_gene_ensembl")
frog       <- biomaRt::useMart("ensembl", dataset = "xtropicalis_gene_ensembl")
coelacanth <- biomaRt::useMart("ensembl", dataset = "lchalumnae_gene_ensembl")
fugu       <- biomaRt::useMart("ensembl", dataset = "ttruncatus_gene_ensembl")
zebrafish  <- biomaRt::useMart("ensembl", dataset = "drerio_gene_ensembl")
#hagfish    <- biomaRt::useMart("ensembl", dataset = "eburgeri_gene_ensembl")
#lamprey    <- biomaRt::useMart("ensembl", dataset = "pmarinus_gene_ensembl")
worm       <- biomaRt::useMart("ensembl", dataset = "celegans_gene_ensembl")
fly        <- biomaRt::useMart("ensembl", dataset = "dmelanogaster_gene_ensembl")
yeast      <- biomaRt::useMart("ensembl", dataset = "scerevisiae_gene_ensembl")

#species_lookup_table <- c("human","chimp","orangutan","rhesus","marmoset","mouse",
#                  "guinea_pig","dog","cow","armadillo","oppossum","platypus",
#                  "chicken","lizard","frog","coelacanth","fugu","zebrafish",
#                  "hagfish","lamprey","worm","fly","yeast")

species_lookup_table <- c("human","chimp","orangutan","rhesus","marmoset","mouse",
                  "guinea_pig","dog","cow","armadillo","oppossum","platypus",
                  "chicken","lizard","frog","coelacanth","fugu","zebrafish",
                  "worm","fly","yeast")

#branch_lookup_table <- c("16","15","14","13","12","11","11","10","10","9","8",
#                         "7","6","6","5","4","3","3","2","2","1","1","0")

branch_lookup_table <- c("15","14","13","12","11","10","10","9","9","8",
                         "7","6","5","5","4","3","2","2","1","1","0")

branch_lengths <- tibble(branch=c(15:1),mya=c(6,6,7,12,18,27,30,60,60,90,60,30,50,350,500))

#dbs <- c(human,chimp,orangutan,rhesus,marmoset,mouse,guinea_pig,dog,cow,
#         armadillo,opossum,platypus,chicken,lizard,frog,coelacanth,fugu,
#         zebrafish,hagfish,lamprey,worm,fly,yeast)

dbs <- c(human,chimp,orangutan,rhesus,marmoset,mouse,guinea_pig,dog,cow,
         armadillo,opossum,platypus,chicken,lizard,frog,coelacanth,fugu,
         zebrafish,worm,fly,yeast)
```

```{r}
retrieve_orhologs <- function(species_lookup_table, branch_lookup_table, dbs, gene_list) {
  out <- tibble::tibble(human_id=character(),
                        ortholog_id=character(),
                        organism=character())
  
  for (i in 2:length(species_lookup_table)) {
      
      result <- biomaRt::getLDS(
        attributes = c("ensembl_gene_id"), 
        filters = "ensembl_gene_id", 
        values = gene_list,
        mart = human,
        attributesL = c("ensembl_gene_id"), 
        martL = dbs[[i]])
      
      result <- dplyr::rename(result,human_id=Gene.stable.ID,ortholog_id=Gene.stable.ID.1) %>%
        dplyr::mutate(organism=species_lookup_table[i],branch=branch_lookup_table[i]) %>%
        tibble::as_tibble()
      
      out <- dplyr::bind_rows(out,result)
  }
  return(out)
}
```

### Conservative set orthologs

```{r}
conservative_orthos <- retrieve_orhologs(species_lookup_table,branch_lookup_table,dbs,conservative_gene_list)

in_humans <- length(unique(conservative_gene_list)) - length(unique(conservative_orthos$human_id))

g <- conservative_orthos %>% 
  dplyr::group_by(human_id) %>% 
  dplyr::summarize(oldest=(min(as.integer(branch)))) %>%
  dplyr::ungroup() %>%
  dplyr::bind_rows(.,tibble::tibble(human_id=as.character(1:in_humans),oldest=rep(16,in_humans))) %>%
  dplyr::group_by(oldest) %>%
  dplyr::summarize(n=n()) %>%
  dplyr::ungroup() %>%
  dplyr::inner_join(branch_lengths,by=c("oldest"="branch")) %>%
  ggplot(aes(x=oldest,y=n/mya)) + geom_bar(stat="identity") +
  xlab("Oldest Branch") + 
  ylab("Events per Million Years") +
  scale_x_continuous(breaks=0:15,labels=c("0","1","2","3","4","5","6","7","8","9","10","11","12","13","14","15"))

print(g)
cowplot::save_plot("../output/aging/conservative_ages.svg",g)
cowplot::save_plot("../output/aging/conservative_ages.png",g)
```

### Standard set orthologs

```{r}
standard_orthos <- retrieve_orhologs(species_lookup_table,branch_lookup_table,dbs,standard_gene_list)

in_humans <- length(unique(standard_gene_list)) - length(unique(standard_orthos$human_id))

g <- standard_orthos %>% 
  dplyr::group_by(human_id) %>% 
  dplyr::summarize(oldest=(min(as.integer(branch)))) %>%
  dplyr::ungroup() %>%
  dplyr::bind_rows(.,tibble::tibble(human_id=as.character(1:in_humans),oldest=rep(16,in_humans))) %>%
  dplyr::group_by(oldest) %>%
  dplyr::summarize(n=n()) %>%
  dplyr::ungroup() %>%
  dplyr::inner_join(branch_lengths,by=c("oldest"="branch")) %>%
  ggplot(aes(x=oldest,y=n/mya)) + geom_bar(stat="identity") +
  xlab("Oldest Branch") + 
  ylab("Events per Million Years") +
  scale_x_continuous(breaks=0:15,labels=c("0","1","2","3","4","5","6","7","8","9","10","11","12","13","14","15"))

print(g)
cowplot::save_plot("../output/aging/standard_ages.svg",g)
cowplot::save_plot("../output/aging/standard_ages.png",g)
```

## Parent-offspring assignments

```{r}
conservative_rbh <- readr::read_tsv("../output/thirdset/conservative_reciprocal_best_hits.tsv",col_names=T)

tmp1 <- conservative_orthos %>% 
  dplyr::group_by(human_id) %>% 
  dplyr::summarize(oldest=(min(as.integer(branch)))) %>%
  dplyr::ungroup() %>%
  dplyr::inner_join(conservative_rbh, by=c("human_id"="query")) %>%
  dplyr::rename(query=human_id,query_branch=oldest)

tmp2 <- conservative_orthos %>% 
  dplyr::group_by(human_id) %>% 
  dplyr::summarize(oldest=(min(as.integer(branch)))) %>%
  dplyr::ungroup() %>%
  dplyr::inner_join(conservative_rbh, by=c("human_id"="subject")) %>%
  dplyr::rename(subject=human_id,subject_branch=oldest)

potbl <- dplyr::inner_join(tmp1,tmp2)
rm(tmp1,tmp2)

potbl$query_relation <- apply(potbl, 1, function(row) {ifelse(row[[2]] == row[[4]],"same",ifelse(row[[2]] < row[[4]],"parent","offspring"))})
potbl$subject_relation <- apply(potbl, 1, function(row) {ifelse(row[[2]] == row[[4]],"same",ifelse(row[[2]] > row[[4]],"parent","offspring"))})

query   <- potbl[,c(1,2,5)] %>% dplyr::rename(gene_id=query,branch=query_branch,relation=query_relation)
subject <- potbl[,c(3,4,6)] %>% dplyr::rename(gene_id=subject,branch=subject_branch,relation=subject_relation)
potbl   <- dplyr::bind_rows(query,subject) %>% dplyr::distinct()

g <- ggplot(potbl,aes(x=branch)) + 
  geom_bar(stat="count") + 
  facet_grid(~relation)

print(g)
 
cowplot::save_plot("../output/aging/parent_offspring.svg",g)
cowplot::save_plot("../output/aging/parent_offspring.png",g)
```

## Tissue expression analysis

### What tissues are these genes expressed in?

```{r}
tpm_threshold <- 2
relationships <- c("same","parent","offspring")
gtex <- readr::read_tsv("../data/GTEx_Analysis_2016-01-15_v7_RNASeQCv1.1.8_gene_median_tpm.gct",skip=2,col_names=T) %>%
  dplyr::select(-Description) %>%
  tidyr::gather(key = tissue, value = exp, -gene_id)

# remove annoying ".#" parts of each gene ID
gtex$gene_id <- sapply(stringr::str_split(gtex$gene_id,"[.]"), function(x) {x[1]})

fgtex <- dplyr::inner_join(gtex,potbl) %>%
  dplyr::distinct()

for (r in relationships) {
  tissue_expression <- fgtex %>%
    dplyr::filter(relation==r) %>%
    dplyr::filter(exp > tpm_threshold) %>%
    dplyr::group_by(tissue) %>%
    dplyr::summarise(n=n()) %>%
    dplyr::arrange(desc(n))
  
  g <- ggplot(tissue_expression, aes(x=reorder(tissue, n),y=n)) +
    geom_bar(stat="identity") +
    coord_flip() +
    ylab("Number of genes") +
    xlab("") + 
    ggtitle(r)
  
  cowplot::save_plot(paste0("../output/aging/",r,"_tissue_expression_barplot.png"),plot=g)
  
  print(g)
}
```

### How many tissues are most genes expressed in?

```{r}
for (relation in relationships) {
  gene_expression <- fgtex %>%
    dplyr::filter(relation==relation) %>%
    dplyr::group_by(gene_id) %>%
    dplyr::summarise(n = sum(exp > tpm_threshold))
  
  g <- ggplot(gene_expression, aes(x=n)) +
    geom_histogram(color="grey70") +
    xlab("Number of Tissues") +
    ylab("Number of Genes")
  
  cowplot::save_plot(paste0("../output/aging/",relation,"_gene_expression_distribution.png"),plot=g)
  
  print(g)
}
```

```{r}
g <- ggplot(fgtex, aes(x=relation,y=log2(exp+1))) +
  geom_boxplot() +
  xlab("Relationship") +
  ylab("Log2 Expression")
  
cowplot::save_plot(paste0("../output/aging/",relation,"_gene_expression_distribution.png"),plot=g)
  
print(g)
```
